### ISSUE TITLE
Compile YOLO and other utralytics models using torch.compile(0

### ISSUE THREAD

**Issue by Abhranta** [ultralytics/ultralytics#6022]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

I am trying to inference yolo and other ultralytics models. I am trying to compile them using torch.compile to see the performace difference. But I am unable to do this. I keep getting the following error: 
 '''
Module [YOLO] is missing the required "forward" function
'''

Can someone help me out ?

### Minimal Reproducible Example

model = YOLO(yolov8x.pt)
model = torch.compile(model)
input_img = 'img.jpg'
output = model(img)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
The underlying PyTorch model is at `model.model`, so you run `torch.compile(model.model)` for example. If you want to compile this and then run it, i.e.:

```python
import torch

from ultralytics import YOLO

model = YOLO()

model.model = torch.compile(model.model)

results = model()

```

**Comment by Abhranta** [comment]
yep , got it working !

**Comment by glenn-jocher** [comment]
@Abhranta that's great to hear! I'm glad you were able to resolve your issue. If you have any other questions, concerns, or run into further challenges with YOLOv8, please don't hesitate to ask. Happy coding!

**Comment by zavinator** [comment]
@glenn-jocher I apologize for the beginner's query, but does the following code snippet look logical? Would executing this with TorchScript in C++ improve its speed?

```python
model.model = torch.compile(model.model)
model.export(format="torchscript")
```

**Comment by glenn-jocher** [comment]
@zavinator hi there,

You're on the right track! Using `torch.compile` is intended to enhance the computational performance of your model. The optimization works by fusing multiple operations in the model into a single operation, leading to faster inference times.

In terms of exporting to TorchScript, it can definitely be beneficial, especially if you're planning on deploying your model in a production environment where Python is not available. TorchScript allows you to serialize your model so it can be loaded and run in a non-Python environment.

Do note that performance improvements can depend on the specifics of your system and the model. Therefore, the real impact of both these methods should ideally be measured and compared with your original setup using a suitable benchmarking process.

As always, it's good to experiment, measure and then make informed decisions. 

Good luck!

**Comment by daniil-lyakhov** [comment]
> The underlying PyTorch model is at `model.model`, so you run `torch.compile(model.model)` for example. If you want to compile this and then run it, i.e.:
> 
> ```python
> import torch
> 
> from ultralytics import YOLO
> 
> model = YOLO()
> 
> model.model = torch.compile(model.model)
> 
> results = model()
> ```

Please be aware that the compiled model is being throw away from the YOLO class instance in case of `.val` method usage. So, it is not possible to validate compiled model by the `YOLO.val`

**Comment by glenn-jocher** [comment]
Thank you for bringing this to our attention. It appears that the issue you're encountering is related to the `AutoBackend` class, which manages the model loading and inference backend dynamically.

To address your concern about the compiled model being discarded during validation with the `YOLO.val` method, you can manually handle the validation process using the compiled model. Here's an example of how you can achieve this:

```python
import torch
from ultralytics import YOLO

# Load and compile the model
model = YOLO('yolov8x.pt')
compiled_model = torch.compile(model.model)

# Manually handle validation
def custom_val(model, data_loader):
    model.eval()
    with torch.no_grad():
        for images, targets in data_loader:
            outputs = model(images)
            # Add your validation logic here

# Example usage with a DataLoader
# Assuming you have a DataLoader named 'val_loader'
# custom_val(compiled_model, val_loader)
```

This approach allows you to bypass the `AutoBackend` and directly use the compiled model for validation.

Regarding the issue with YOLOv8 and YOLOv10 not working correctly after model compilation, we are aware of the ongoing discussion in [issue #13800](https://github.com/ultralytics/ultralytics/issues/13800). Our team is actively investigating this, and we appreciate your patience.

If you encounter any further issues or have additional questions, please feel free to reach out. Your feedback is invaluable in helping us improve our models and tools. Thank you for your understanding and support!

**Comment by Ilyabasharov** [comment]
@daniil-lyakhov Hello, did you solve this issue?

**Comment by daniil-lyakhov** [comment]
> @daniil-lyakhov Hello, did you solve this issue?

Hi, nothing has changed as far as I know

**Comment by caiodataopshouse** [comment]
@zavinator Hi. For me this piece of code don't make much sense. torch.compile() will improve performance on the current runtime.. It will not influence anything if you compile again using torchscript.

**Comment by CongSuxu** [comment]
Hi, I’m trying to use `torch.compile` to accelerate inference with the Ultralytics YOLO model.

I noticed that if I compile the model like this:

```python
from ultralytics import YOLO
import torch

model = YOLO()
model.model = torch.compile(model.model)
```
It doesn't actually run the compiled model during inference, because internally model.predict() executes:

```python
if fuse:
    weights = weights.fuse(verbose=verbose) # In AutoBackend
```

This replaces the compiled weights (OptimizedModule) with a new DetectionModel, which removes the benefits of torch.compile.

To ensure the compiled model is actually used during inference, I tried this:

```python
model.predictor.model.model = torch.compile(model.predictor.model.model)
# YOLO.DetectionPredictor(BasePredictor).AutoBackend.DetectionModel
```

This successfully sets the final running model to an OptimizedModule.

However, when running inference, I get lots of recompile warnings, like:

```pgsql
V0704 15:58:07.714259 9896 torch\_dynamo\guards.py:2611] [3/2] [__recompiles] Recompiling function forward_fuse in C:\Users\admin\Desktop\torchServe\system\python\lib\site-packages\ultralytics\nn\modules\conv.py:81
V0704 15:58:07.714259 9896 torch\_dynamo\guards.py:2611] [3/2] [__recompiles]     triggered by the following guard failure(s):
V0704 15:58:07.714259 9896 torch\_dynamo\guards.py:2611] [3/2] [__recompiles]     - tensor 'L['x']' stride mismatch at index 0. expected 3145728, actual 8388608
V0704 15:58:07.973715 9332 torch\_dynamo\guards.py:2611] [3/3] [__recompiles] Recompiling function forward_fuse in C:\Users\admin\Desktop\torchServe\system\python\lib\site-packages\ultralytics\nn\modules\conv.py:81
V0704 15:58:07.973715 9332 torch\_dynamo\guards.py:2611] [3/3] [__recompiles]     triggered by the following guard failure(s):
V0704 15:58:07.973715 9332 torch\_dynamo\guards.py:2611] [3/3] [__recompiles]     - tensor 'L['x']' stride mismatch at index 0. expected 3145728, actual 8388608
V0704 15:58:07.973715 9332 torch\_dynamo\guards.py:2611] [3/3] [__recompiles]     - tensor 'L['x']' Tensor device index mismatch. Expected device index to be , actual  
V0704 15:58:09.023635 9896 torch\_dynamo\guards.py:2611] [3/4] [__recompiles] Recompiling function forward_fuse in C:\Users\admin\Desktop\torchServe\system\python\lib\site-packages\ultralytics\nn\modules\conv.py:81
V0704 15:58:09.023635 9896 torch\_dynamo\guards.py:2611] [3/4] [__recompiles]     triggered by the following guard failure(s):
V0704 15:58:09.023635 9896 torch\_dynamo\guards.py:2611] [3/4] [__recompiles]     - tensor 'L['x']' Tensor device index mismatch. Expected device index to be  , actual 
V0704 15:58:09.023635 9896 torch\_dynamo\guards.py:2611] [3/4] [__recompiles]     - tensor 'L['x']' size mismatch at index 1. expected 32, actual 128
V0704 15:58:09.023635 9896 torch\_dynamo\guards.py:2611] [3/4] [__recompiles]     - tensor 'L['x']' Tensor device index mismatch. Expected device index to be  , actual 
V0704 15:58:09.023635 9896 torch\_dynamo\guards.py:2611] [3/4] [__recompiles]     - tensor 'L['x']' stride mismatch at index 0. expected 3145728, actual 8388608
...
W0704 15:25:10.276125 22712 torch\_dynamo\convert_frame.py:762] [3/8] torch._dynamo hit config.cache_size_limit (8)
W0704 15:25:10.276125 22712 torch\_dynamo\convert_frame.py:762] [3/8]    function: 'forward_fuse' (D:\torchServe\system\python\lib\site-packages\ultralytics\nn\modules\conv.py:81)
W0704 15:25:10.276125 22712 torch\_dynamo\convert_frame.py:762] [3/8]    last reason: tensor 'L['x']' stride mismatch at index 0. expected 540672, actual 8126464
W0704 15:25:10.276125 22712 torch\_dynamo\convert_frame.py:762] [3/8] To log all recompilation reasons, use TORCH_LOGS="recompiles".
W0704 15:25:10.276125 22712 torch\_dynamo\convert_frame.py:762] [3/8] To diagnose recompilation issues, see https://pytorch.org/docs/main/torch.compiler_troubleshooting.html.
```
Even though I only use a single input shape, and all preprocessing outputs are fixed-size tensors, recompilation still occurs frequently — especially inside this function:

```bash
ultralytics/nn/modules/conv.py:81 -> forward_fuse
```
Despite the recompilation warnings, the program still runs successfully after a very slow first batch (around 35 seconds).

**Comment by glenn-jocher** [comment]
@CongSuxu thanks for the detailed analysis! You've identified a key challenge with `torch.compile` and the YOLO inference pipeline. The recompilations are occurring because torch.compile optimizes for specific tensor shapes and strides, but the dynamic nature of image preprocessing and the model's internal operations can produce varying tensor characteristics.

The issue with the `fuse` operation replacing your compiled model is a known limitation - the AutoBackend recreates the model during inference preparation, which indeed discards the compiled version.

Your approach of compiling `model.predictor.model.model` is on the right track, but the recompilations suggest that torch.compile is encountering different tensor shapes/strides than expected. This often happens when:

1. Input images have different dimensions even after preprocessing
2. The model's internal operations create tensors with varying memory layouts
3. Device transfers occur during inference

For now, the most stable approach for using torch.compile with YOLO models is to work directly with the underlying model outside of the full YOLO inference pipeline, similar to what was suggested earlier in this thread. This avoids the AutoBackend complexities but requires manual handling of preprocessing and postprocessing.

The torch.compile integration with the full YOLO pipeline is still being refined, and your feedback helps identify these edge cases.

**Comment by UltralyticsAssistant** [comment]
Great news — a potential fix for this issue has been merged in PR [ultralytics/ultralytics#21975](https://github.com/ultralytics/ultralytics/pull/21975), authored by @glenn-jocher with contributions from @ExtReMLapin and @Y-T-G.

What changed and why it may help:
- Optional PyTorch 2.x `torch.compile` added across train/val/predict for speedups, with safe fallback if unsupported.
- New `unwrap_model` utility replaces `de_parallel`, working for both compiled and DataParallel models to improve stability across EMA, TensorBoard, and validators.
- Safer DataLoader handling with `drop_last=True` when compiling to keep static shapes and reduce recompilations.
- Validation during training uses the original (non-compiled) model for reliability; final eval explicitly avoids compile for consistency.
- Minor fixes and cleanups that address shape edge cases and logging.

How to test the fix:
- Upgrade to the version containing the changes:
  - `pip install -U ultralytics>=8.3.196`
- Or install the latest main:
  - `pip install git+https://github.com/ultralytics/ultralytics.git@main`
- Usage examples:
  - CLI:
    - `yolo train model=yolo11n.pt data=coco8.yaml epochs=50 compile=True`
  - Python:
    ```python
    from ultralytics import YOLO

    model = YOLO("yolo11n.pt")
    model.train(compile=True)
    # optionally:
    model.val(compile=True)
    model.predict("img.jpg", compile=True)
    ```

Notes:
- Requires PyTorch 2.x; if unsupported, it will gracefully fall back to eager mode.
- First run may take longer due to compile time; subsequent iterations should be faster.

Please try the update and share whether this resolves the issue. Thanks for reporting it 🙏 and feel free to provide any additional details if the problem persists.
